### PR TITLE
AI can now use vending machines

### PR DESCRIPTION
**Pull request by Simplehorror**
# Document the changes in your pull request

It feels strange that the ship computer can't dispense a cola. Now you can ask the AI to dispense that life saving medical spray instead of waiting for your paycheck. Wonder if this could be integrated with corporate? :P

# Changelog

:cl:  
tweak: The AI has finally finished reverse engineering the vending machine subsystems
/:cl:

### CODE REVIEW

**Comment by Bibby0110** [comment]
Make it an AI upgrade

**Comment by Simplehorror** [comment]
> Make it an AI upgrade

Good idea, gimmie a second

For anyone concerned about the AI being abused to dispense items for free; it's no different than the AI being abused for access and is mostly a security issue. It's also an indication that the economy is unbalanced if the AI is constantly being used to dispense, and I was never a big fan of it.

**Comment by bruhlookatthisdood** [comment]
Hear me out (have fun trying to make this work)
If AI dispenses from med vendor, subtracts the cost from med budget
Dispensing from science vendors, Sci budget, ect.
Seems like a better solution than to just make everything free for them, especcialy things like the advanced medkits in medical lockers
If you dont I see a possible abuse of the corprat lawset where they will just dispense everything from the vendors round start to "save costs" so no one has to waste money.

**Comment by Simplehorror** [comment]
> Hear me out (have fun trying to make this work) If AI dispenses from med vendor, subtracts the cost from med budget Dispensing from science vendors, Sci budget, ect.

If I do this it might be in another PR because I'll have to edit tgui because the AI doesn't have a card and vending machines stupidly want you to have a card

**Comment by LoliconSlayer** [comment]
Who?

File: code/modules/mob/living/silicon/ai/decentralized/projects/vending_machine.dm [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/datum/ai_project/vending_machine
+	name = "Unlock vending machines"
+	description = "By reverse engineering the security subsystems in the vending machines, we can bypass the access and payment gates and dispense items."
+	research_cost = 500
+	ram_required = 1
```
Comment (Bibby0110) [inline code/modules/mob/living/silicon/ai/decentralized/projects/vending_machine.dm]:
```suggestion
	can_be_run = FALSE
```

File: code/modules/mob/living/silicon/ai/decentralized/projects/vending_machine.dm [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/datum/ai_project/vending_machine
+	name = "Unlock vending machines"
```
Comment (Bibby0110) [inline code/modules/mob/living/silicon/ai/decentralized/projects/vending_machine.dm]: Vending Machine Control

File: code/modules/mob/living/silicon/ai/decentralized/projects/vending_machine.dm [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+	description = "By reverse engineering the security subsystems in the vending machines, we can bypass the access and payment gates and dispense items."
+	research_cost = 500
+	ram_required = 1
+	research_requirements_text = "None"
+	category = AI_PROJECT_MISC
+
+/datum/ai_project/vending_machine/run_project(force_run = FALSE)
```
Comment (Bibby0110) [inline code/modules/mob/living/silicon/ai/decentralized/projects/vending_machine.dm]:
```suggestion
/datum/ai_project/vending_machine/finish()
```

File: code/modules/mob/living/silicon/ai/decentralized/projects/vending_machine.dm [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+	. = ..()
+	if(!.)
+		return .
+	ai.ignores_capitalism = TRUE
+
+
+/datum/ai_project/vending_machine/stop()
```
Comment (Bibby0110) [inline code/modules/mob/living/silicon/ai/decentralized/projects/vending_machine.dm]: remove this

File: code/modules/mob/living/silicon/ai/decentralized/projects/vending_machine.dm [hunk 5]
```diff
@@ -0,0 +1,18 @@
+/datum/ai_project/vending_machine
+	name = "Unlock vending machines"
+	description = "By reverse engineering the security subsystems in the vending machines, we can bypass the access and payment gates and dispense items."
+	research_cost = 500
+	ram_required = 1
+	research_requirements_text = "None"
+	category = AI_PROJECT_MISC
+
+/datum/ai_project/vending_machine/run_project(force_run = FALSE)
+	. = ..()
```
Comment (Bibby0110) [inline code/modules/mob/living/silicon/ai/decentralized/projects/vending_machine.dm]: remove everything but ai.ignores_capitalism = TRUE

**Review by Bibby0110 — CHANGES_REQUESTED** [review]
This turns it into a passive upgrade (No RAM usage) which seems more fair.
If you don't want it to be passive just do the name change

**Comment by Simplehorror** [comment]
Done

**Review by Bibby0110 — APPROVED** [review]

**Review by SomeguyManperson — CHANGES_REQUESTED** [review]
At the very least make sure this won't allow the AI to buy premium items since those are generally balanced around the people who need them actually needing to spend money on them
Otherwise people will roundstart "law 2 ai dump the expensive shit out of the sec/med vendor"

**Comment by Simplehorror** [comment]
> At the very least make sure this won't allow the AI to buy premium items since those are generally balanced around the people who need them actually needing to spend money on them Otherwise people will roundstart "law 2 ai dump the expensive shit out of the sec/med vendor"

security/admin issue, same issue with AI LAW 2 OPEN TO SPARE

**Comment by ToasterBiome** [comment]
> > At the very least make sure this won't allow the AI to buy premium items since those are generally balanced around the people who need them actually needing to spend money on them Otherwise people will roundstart "law 2 ai dump the expensive shit out of the sec/med vendor"
> 
> security/admin issue, same issue with AI LAW 2 OPEN TO SPARE

why would we want to code more security/admin issues into the game?

**Comment by bruhlookatthisdood** [comment]
still suggesting the "anything the ai buys will be taken from the department's budget" idea

**Comment by Simplehorror** [comment]
> > > At the very least make sure this won't allow the AI to buy premium items since those are generally balanced around the people who need them actually needing to spend money on them Otherwise people will roundstart "law 2 ai dump the expensive shit out of the sec/med vendor"
> > 
> > 
> > security/admin issue, same issue with AI LAW 2 OPEN TO SPARE
> 
> why would we want to code more security/admin issues into the game?

every line of code in this game is an admin issue its the reason we have so many, people will find a way to abuse anything. this will give more benefits to the crew than negatives

**Comment by JamieD1** [comment]
> At the very least make sure this won't allow the AI to buy premium items since those are generally balanced around the people who need them actually needing to spend money on them Otherwise people will roundstart "law 2 ai dump the expensive shit out of the sec/med vendor"

Do this

**Comment by JamieD1** [comment]
> still suggesting the "anything the ai buys will be taken from the department's budget" idea

Also do this.

**Comment by bruhlookatthisdood** [comment]
maybe there can be two upgrades?
one is basic vendors and the other is advanced vendors, basic vendors giving access to the smoking machines, soda vendors, ect. and advanced vendors giving access to medical, security, ect.

**Comment by Bibby0110** [comment]
> > still suggesting the "anything the ai buys will be taken from the department's budget" idea
> 
> Also do this.

Doesn't actually matter
Budgets will always have enough money
(Bar maybe the sec budget on revs/clockies)

**Comment by bruhlookatthisdood** [comment]
> > > still suggesting the "anything the ai buys will be taken from the department's budget" idea
> > 
> > 
> > Also do this.
> 
> Doesn't actually matter Budgets will always have enough money (Bar maybe the sec budget on revs/clockies)

give the AI it's own personal bank account

**Comment by Simplehorror** [comment]
> > At the very least make sure this won't allow the AI to buy premium items since those are generally balanced around the people who need them actually needing to spend money on them Otherwise people will roundstart "law 2 ai dump the expensive shit out of the sec/med vendor"
> 
> Do this
premium items being the ones that cost money even for department members?

**Comment by Simplehorror** [comment]
> > > > still suggesting the "anything the ai buys will be taken from the department's budget" idea
> > > 
> > > 
> > > Also do this.
> > 
> > 
> > Doesn't actually matter Budgets will always have enough money (Bar maybe the sec budget on revs/clockies)
> 
> give the AI it's own personal bank account

could be funny and incentivise AI as well as giving them a new lawset (maximise your bank)

**Comment by bruhlookatthisdood** [comment]
> > > > > still suggesting the "anything the ai buys will be taken from the department's budget" idea
> > > > 
> > > > 
> > > > Also do this.
> > > 
> > > 
> > > Doesn't actually matter Budgets will always have enough money (Bar maybe the sec budget on revs/clockies)
> > 
> > 
> > give the AI it's own personal bank account
> 
> could be funny and incentivise AI as well as giving them a new lawset (maximise your bank)

Corporate Greed lawset

**Comment by JamieD1** [comment]
How we coming along with this?

**Pull request closed without merging**